File: 0116. Populating Next Right Pointers in Each Node/Solution0116.cpp
class Solution0116 {
public:
    Node* connect(Node* root) {
        if (root == NULL) {
            return root;
        }
        Node* leftmost = root;
        while (leftmost->left != NULL) {
            Node* cur = leftmost;
            while (cur != NULL) {
                cur->left->next = cur->right;
                if (cur->next != NULL) {
                    cur->right->next = cur->next->left;
                }
                cur = cur->next;
            }
            leftmost = leftmost->left;
        }
        return root;
    }
};

File: 0116. Populating Next Right Pointers in Each Node/Solution0116.py
class Solution0116:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root
        leftmost = root
        while leftmost.left:
            cur = leftmost
            while cur:
                cur.left.next = cur.right
                if cur.next:
                    cur.right.next = cur.next.left
                cur = cur.next
            leftmost = leftmost.left
        return root